modules/xmpp/moderator.js

```diff
@@ -39,549 +41,639 @@ function createExpBackoffTimer(step) {
+    constructor(xmpp) {
+        super();
 
-    // External authentication stuff
-    this.externalAuthEnabled = false;
+        this.getNextTimeout = createExpBackoffTimer(1000);
+        this.getNextErrorTimeout = createExpBackoffTimer(1000);
+        this.options = xmpp.options;
 
-    // Whether SIP gateway (jigasi) support is enabled. TODO: use presence so it can be changed based on jigasi
-    // availability.
-    this.sipGatewayEnabled = false;
+        // External authentication stuff
+        this.externalAuthEnabled = false;
 
-    this.eventEmitter = emitter;
+        // Whether SIP gateway (jigasi) support is enabled. TODO: use presence so it can be changed based on jigasi
+        // availability.
+        this.sipGatewayEnabled = false;
 
-    this.connection = xmpp.connection;
+        this.xmpp = xmpp;
+        this.connection = xmpp.connection;
 
-    // The JID to which conference-iq requests are sent over XMPP.
-    this.targetJid = this.options.hosts?.focus;
+        // The JID to which conference-iq requests are sent over XMPP.
+        this.targetJid = this.options.hosts?.focus;
 
-    // If not specified default to 'focus.domain'
-    if (!this.targetJid) {
-        this.targetJid = `focus.${this.options.hosts?.domain}`;
-    }
+        // If not specified default to 'focus.domain'
+        if (!this.targetJid) {
+            this.targetJid = `focus.${this.options.hosts?.domain}`;
+        }
 
-    this.targetUrl = this.options.conferenceRequestUrl;
+        this.targetUrl = this.options.conferenceRequestUrl;
 
-    // Whether to send conference requests over HTTP or XMPP
-    this.mode = this.targetUrl ? 'http' : 'xmpp';
-    logger.info(`Using ${this.mode} for conference requests.`);
+        // Whether to send conference requests over HTTP or XMPP
+        this.mode = this.targetUrl ? 'http' : 'xmpp';
+        logger.info(`Using ${this.mode} for conference requests.`);
 
-    // The set of JIDs known to belong to jicofo. Populated from configuration
-    // and responses from conference requests.
-    this.focusUserJids = new Set();
+        // The set of JIDs known to belong to jicofo. Populated from configuration
+        // and responses from conference requests.
+        this.focusUserJids = new Set();
 
-    if (options.focusUserJid) {
-        this.focusUserJids.add(options.focusUserJid);
-    }
+        if (this.options.focusUserJid) {
+            this.focusUserJids.add(this.options.focusUserJid);
+        }
 
-    // FIXME: Message listener that talks to POPUP window
-    /**
-     *
-     * @param event
-     */
-    function listener(event) {
-        if (event.data && event.data.sessionId) {
-            if (event.origin !== window.location.origin) {
-                logger.warn(`Ignoring sessionId from different origin: ${event.origin}`);
+        // FIXME: Message listener that talks to POPUP window
+        /**
+         *
+         * @param event
+         */
+        function listener(event) {
```

Comment:
I think this function can be outside of the constructor. WDYT?

Reply:
Well I have kept the code as it is. We can improve it later.